src/librustc_codegen_ssa/base.rs

```diff
@@ -508,6 +509,43 @@ fn get_argc_argv<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
 
 pub const CODEGEN_WORKER_ID: usize = ::std::usize::MAX;
 
+/// check for the #[rustc_error] annotation, which forces an
+/// error in codegen. This is used to write compile-fail tests
+/// that actually test that compilation succeeds without
+/// reporting an error.
+pub fn check_for_rustc_errors_attr(tcx: TyCtxt<'_>) {
+    if let Some((def_id, _)) = tcx.entry_fn(LOCAL_CRATE) {
+        let attrs = &*tcx.get_attrs(def_id);
+        for attr in attrs {
+            if attr.check_name(sym::rustc_error) {
```

Comment:
...and using `.filter(...)` here on `get_attrs(def_id).iter()` to reduce rightward drift some more.